File: LinkedList/python/problem_083_remove_duplicates.py
"""
Problem 83: Remove Duplicates from Sorted List
Difficulty: Easy

Problem Statement:
Given the head of a sorted linked list, delete all duplicates such that each element appears only once.

Example:
Input: head = [1,1,2,3,3]
Output: [1,2,3]

Time Complexity: O(n)
Space Complexity: O(1)
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        current = head
        
        while current and current.next:
            if current.val == current.next.val:
                # Skip duplicate
                current.next = current.next.next
            else:
                # Move to next distinct value
                current = current.next
        
        return head

def create_list(arr):
    if not arr:
        return None
    head = ListNode(arr[0])
    current = head
    for i in range(1, len(arr)):
        current.next = ListNode(arr[i])
        current = current.next
    return head

def print_list(head):
    result = []
    current = head
    while current:
        result.append(str(current.val))
        current = current.next
    print("[" + ",".join(result) + "]")

if __name__ == "__main__":
    solution = Solution()
    head = create_list([1, 1, 2, 3, 3])
    print("Original: ", end="")
    print_list(head)
    result = solution.deleteDuplicates(head)
    print("After removing duplicates: ", end="")
    print_list(result)

File: LinkedList/java/Problem083_RemoveDuplicates.java
/**
 * Problem 83: Remove Duplicates from Sorted List
 * Difficulty: Easy
 * 
 * Problem Statement:
 * Given the head of a sorted linked list, delete all duplicates such that each element appears only once.
 * 
 * Example:
 * Input: head = [1,1,2,3,3]
 * Output: [1,2,3]
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
}

public class Problem083_RemoveDuplicates {
    
    public ListNode deleteDuplicates(ListNode head) {
        ListNode current = head;
        
        while (current != null && current.next != null) {
            if (current.val == current.next.val) {
                // Skip duplicate
                current.next = current.next.next;
            } else {
                // Move to next distinct value
                current = current.next;
            }
        }
        
        return head;
    }
    
    public static ListNode createList(int[] arr) {
        if (arr.length == 0) return null;
        ListNode head = new ListNode(arr[0]);
        ListNode current = head;
        for (int i = 1; i < arr.length; i++) {
            current.next = new ListNode(arr[i]);
            current = current.next;
        }
        return head;
    }
    
    public static void printList(ListNode head) {
        ListNode current = head;
        System.out.print("[");
        while (current != null) {
            System.out.print(current.val);
            if (current.next != null) System.out.print(",");
            current = current.next;
        }
        System.out.println("]");
    }
    
    public static void main(String[] args) {
        Problem083_RemoveDuplicates solution = new Problem083_RemoveDuplicates();
        ListNode head = createList(new int[]{1, 1, 2, 3, 3});
        System.out.print("Original: ");
        printList(head);
        ListNode result = solution.deleteDuplicates(head);
        System.out.print("After removing duplicates: ");
        printList(result);
    }
}
